src/java_bytecode/java_bytecode_convert_class.cpp

```diff
@@ -145,10 +197,28 @@ void java_bytecode_convert_classt::convert(const classt &c)
 
   if(!c.extends.empty())
   {
-    symbol_typet base("java::"+id2string(c.extends));
-    class_type.add_base(base);
+    const symbol_typet base("java::" + id2string(c.extends));
+
+    // if the superclass is generic then the class has the superclass reference
+    // including the generic info in its signature
+    // e.g., signature for class 'A<T>' that extends
+    // 'Generic<Integer>' is '<T:Ljava/lang/Object;>LGeneric<LInteger;>;'
+    if(
+      c.signature.has_value() &&
+      !extract_generic_superclass_reference(c.signature.value()).empty())
+    {
+      const java_generic_symbol_typet gen_base(
+        base,
+        extract_generic_superclass_reference(c.signature.value()),
+        qualified_classname);
+      class_type.add_base(gen_base);
+    }
+    else
+    {
+      class_type.add_base(base);
+    }
     class_typet::componentt base_class_field;
-    base_class_field.type()=base;
+    base_class_field.type() = class_type.bases().at(0).type();
```

Comment:
How did this work if `base` was an `exprt` and `type()` was a `typet`?

Reply:
`base` was a `symbol_typet`, only once called `.add_base(base)` it was wrapped into an `exprt`.